presto-spi/src/main/java/com/facebook/presto/spi/block/BlockFlattener.java

```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.api.Experimental;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.spi.block.ClosingBlockLease.newLease;
+import static java.util.Objects.requireNonNull;
+
+@Experimental
+public class BlockFlattener
+{
+    private final ArrayAllocator allocator;
+
+    public BlockFlattener(ArrayAllocator allocator)
+    {
+        this.allocator = requireNonNull(allocator, "allocator is null");
+    }
+
+    public BlockLease flatten(Block block)
+    {
+        requireNonNull(block, "block is null");
+        if (block instanceof DictionaryBlock) {
+            return flattenDictionary((DictionaryBlock) block);
+        }
+        if (block instanceof RunLengthEncodedBlock) {
+            return flattenRLE((RunLengthEncodedBlock) block);
+        }
+        return newLease(block);
+    }
+
+    private BlockLease flattenDictionary(DictionaryBlock dictionaryBlock)
+    {
+        Block block = dictionaryBlock.getDictionary();
+        int[] remappedIds = (int[]) dictionaryBlock.getIds().getBase();
+        int positionCount = dictionaryBlock.getPositionCount();
+        int[] leasedIds = null;
+        while (true) {
+            if (block instanceof DictionaryBlock) {
+                dictionaryBlock = (DictionaryBlock) block;
+                int[] ids = (int[]) dictionaryBlock.getIds().getBase();
+
+                if (leasedIds == null) {
+                    leasedIds = allocator.borrowIntArray(positionCount);
+                }
+
+                for (int i = 0; i < positionCount; ++i) {
+                    leasedIds[i] = ids[remappedIds[i] + dictionaryBlock.getOffsetBase()];
+                }
+
+                remappedIds = leasedIds;
+                block = dictionaryBlock.getDictionary();
+            }
+            else if (block instanceof RunLengthEncodedBlock) {
+                RunLengthEncodedBlock rle = (RunLengthEncodedBlock) block;
+                int[] ids = allocator.borrowIntArray(positionCount);
+                Arrays.fill(ids, 0, positionCount, 0);
+                if (leasedIds != null) {
+                    allocator.returnArray(leasedIds);
+                }
+                remappedIds = ids;
+                leasedIds = ids;
+
+                block = rle.getValue();
+            }
+            else {
+                if (leasedIds == null) {
+                    return newLease(dictionaryBlock);
+                }
+
+                block = new DictionaryBlock(positionCount, block, remappedIds);
+                int[] leasedMapToReturn = leasedIds; // effectively final
+                return newLease(block, () -> allocator.returnArray(leasedMapToReturn));
+            }
+        }
+    }
+
+    private BlockLease flattenRLE(RunLengthEncodedBlock rleBLock)
```

Comment:
nit: this method can be static

Reply:
Since the dictionary version can't be, I'd prefer to keep it as is